### PR TITLE
LV2 Fixups

### LINKED ISSUE #12480: [Regression] GTAV Freezing
GTAV is freezing within 15 seconds of starting the game. Possibly related to the swap character and view character stats feature because trying to look at that seemed to act as a trigger for me if at first it wouldn't freeze.

### Build with regression

v0.0.23-14027-2ec03936

### Attach two log files

[Broken.log.gz](https://github.com/RPCS3/rpcs3/files/9277868/RPCS3.log.gz)
[Working.log.gz](https://github.com/RPCS3/rpcs3/files/9277871/RPCS3.log.gz)

### System configuration

Ryzen 5 3600
GTX 1660 Super
Windows 11 Home

### Other details

I was playing as Trevor preparing for Merryweather heist outside Floyd's safehouse and the game wouldn't even let me move at all without freezing.

### LINKED ISSUE #12485: [Regression] Infamous 2
### Quick summary

Infamous 2 not go ingame, freeze on load screen

### Details

_No response_

### Build with regression

0.0.23-14027

### Attach two log files

Log 0.0.23-14027
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/9281265/RPCS3.log)

Log 0.0.23-14014

[RPCS3.log](https://github.com/RPCS3/rpcs3/files/9281264/RPCS3.log)

### Attach capture files for visual issues

_No response_

### System configuration

_No response_

### Other details

_No response_

### LINKED ISSUE #12486: [Regression] Motorstorm Apocalypse Crashes at Boot
### Quick summary

Motorstorm Apocalypse crashes while booting up at around the Evo Studios logo.

### Details

Motorstorm Apocalypse crashes while booting up at around the Evo Studios logo. Tested game version is BCES00484. Last working build is 0.0.23-14014.

### Build with regression

0.0.23-14027

### Attach two log files

(Regression)[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/9281259/RPCS3.log.gz)
(Working)[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/9281274/RPCS3.log.gz)

### Attach capture files for visual issues

_No response_

### System configuration

CPU: Intel i9 9900k (HT Disabled)
GPU: Nvidia Geforce RTX 2060 516.59
OS: Windows 10 Pro 21H2

### Other details

_No response_

### LINKED ISSUE #12489: [Regression] GoW freezing (crash) frequently.
### Quick summary

God of War 1 from the collection cd used to work fine but on builds after [#12378](https://github.com/RPCS3/rpcs3/pull/12378) it freezes so often sometimes I cannot make it past the main menu.

### Details

The game randomly completely locks up, almost always within minutes, sometimes much faster.

### Build with regression

#12378

### Attach two log files

[RPCS3.log (broken)](https://github.com/RPCS3/rpcs3/files/9286115/RPCS3.log)

[RPCS3.log (working)](https://github.com/RPCS3/rpcs3/files/9286149/RPCS3.log)

### Attach capture files for visual issues

_No response_

### System configuration

Operating System
	Windows 11 Pro Insider Preview 64-bit
CPU
	AMD Ryzen Threadripper 3960X
	Castle Peak 7nm Technology
RAM
	64.0GB Unknown @ 1177MHz (15-15-15-35)
Motherboard
	Gigabyte Technology Co. Ltd. TRX40 AORUS PRO WIFI (SP3r2)	
Graphics
	Acer CB281HK (3840x2160@60Hz)
	LG ULTRAWIDE (2560x1080@60Hz)
	Acer V223W (1680x1050@59Hz)
	3071MB NVIDIA GeForce GTX 1080 Ti (ZOTAC International)	
Storage
	223GB ZOTAC ZTSSD-A5P-240G-PE (SATA (SSD))	
	465GB SanDisk Ultra II 500GB (SATA (SSD))	
	931GB Western Digital WDC WD10EADS-00M2B0 (SATA )	
	2794GB TOSHIBA HDWD130 (SATA )	
	931GB Samsung SSD 970 EVO 1TB (Unknown (SSD))
	29GB Lexar USB Flash Drive USB Device (USB )
Audio
	Realtek USB Audio

### Other details

The freeze happens randomly but never takes more than like ~10min for the game to hang.

### PR DESCRIPTION
**Pull request by elad335**
* Optimize lv2_obj::yield, fix a bug that existed before the pr and some which were introduced by it.
* A hotfix in sys_lwmutex/mutex.
* Fix a race in savestates with sys_lwcond.
* Improve the accuracy of LV2 timers regardless of the setting used. (count time passed from the time of sleep, not of wait_timeout)
* ~~**_Fix a race condition which existed before the pr as well. For me it fixes a freeze in Demon Souls after a while but it improves overall PPU stability in nearly every game. description is in the commit: https://github.com/RPCS3/rpcs3/pull/12482/commits/bf4791f571c9ea7be59fbc9b958ae73731ab0672_**~~ Removed from this pr.

Fixes #12480 
Fixes #12485
Fixes #12486
Fixes #12489

### CODE REVIEW

**Comment by Kravickas** [comment]
Gran Turismo 6 broken after #12378 and still broken with PR, I tried 2,4,6,8 PPU threads - no sound, pad doesnt work

```·F 0:00:56.622620 {PPU[0x100001f] Thread (9ZWAM) [0x00c45708]} SIG: Thread terminated due to fatal error: Verification failed
(in file C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\rpcs3\Emu\Cell\lv2\sys_cond.cpp:455[:2], in function sys_cond_wait)
```
PR [RPCS3.log](https://github.com/RPCS3/rpcs3/files/9281412/RPCS3.log)
Master before #12378 [RPCS3.log](https://github.com/RPCS3/rpcs3/files/9281428/RPCS3.log)

**Comment by elad335** [comment]
Test now.

**Comment by Kravickas** [comment]
[e7e3094](https://github.com/RPCS3/rpcs3/commit/e7e3094e28b9fd70751997d7d84a247017b72307) GT6 back ingame

**Comment by coolllman** [comment]
Infamous 2, 2 ppu not working, 4 ppu working

**Comment by elad335** [comment]
@coolllman Retest.

**Comment by coolllman** [comment]
No luck
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/9282591/RPCS3.log)

**Comment by elad335** [comment]
@coolllman The commit has in the log is e7e3094e, can you reupload log with the latest build? (fe302cb)

**Comment by coolllman** [comment]
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/9283091/RPCS3.log)

**Comment by coolllman** [comment]
All works with last changes, thank you.

**Comment by coolllman** [comment]
Motorstorm Apocalypse same freeze on boot

**Comment by elad335** [comment]
L-O-G

**Comment by ERNSTEROCK** [comment]
im having the same issue with infamous 2 I cleared the cache and changed the ppu to 4 and the game was working for a bit then crashed and now it wont get past the loading screen please help.

**Comment by elad335** [comment]
@coolllman Fixed.

**Comment by coolllman** [comment]
Yes, infamous 2 and motorstorm fixed

**Comment by elad335** [comment]
Please test for performance regressions on both Windows and Linux.

**Comment by tech2c** [comment]
Diablo 3 working on Linux. In fact I played for 1 hour 20 mins before the usual VM:Access violation error (unrelated to this PR). It's the longest play-through without a crash so far!

**Comment by elad335** [comment]
Just fixed a e which an annoying race condition which RPCS3 had since LV2 scheduler got implmeneted (many, many years ago). For me it fixes freezes in Demon Souls after a while but it affects every game.

**Comment by Nerboruto** [comment]
is it necessary to clear ppu cache?

**Comment by elad335** [comment]
no.

**Comment by ERNSTEROCK** [comment]
how do I fix infamous 2?

**Comment by elad335** [comment]
Retest.

**Comment by Darkhost1999** [comment]
GTA V regression still fixed. Still an unknown cause of freezing in water as this log does contain but that was happening before PR 12470. And not result of these changes.
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/9293491/RPCS3.log.gz)

**Comment by coolllman** [comment]
Infamous 2 work, motorstorm apocaplypse freeze
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/9293786/RPCS3.log)

**Comment by coolllman** [comment]
No luck
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/9294302/RPCS3.log)

**Comment by elad335** [comment]
@coolllman can you log kernel explorer state when it freezes?

**Comment by coolllman** [comment]
Clear directory don't help, how do rhe logkernel explorer state?

**Comment by elad335** [comment]
1. Pause the emulation (Ctrl+P).
2. Open **Kernel Explorer** in **Utilities** tab on the top menu bar in RPCS3's window and select **Log All**.

**Comment by coolllman** [comment]
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/9297315/RPCS3.log)

**Comment by elad335** [comment]
Well, works for me, maybe try to leaving it running more than a minutte when loading. I think that's it for this pr.
image

**Comment by coolllman** [comment]
Wait 5 minutes, no change with new pr, stuck on the warning screen

**Comment by tech2c** [comment]
I just tried Diablo 3 again with latest PR. It crashed after 5 mins. This is back to normal behavior like master.

**Comment by elad335** [comment]
@coolllman Retest, upload log and log kernel explorer again.

**Comment by coolllman** [comment]
Sorry didn't test earlier pr, on last change motorstorm work

**Comment by Kravickas** [comment]
Tiger 14 and GT6 seems ok

File: rpcs3/rpcs3qt/debugger_frame.cpp [hunk 1]
```diff
@@ -857,20 +858,43 @@ void debugger_frame::UpdateUnitList()
...
+		}
+
+		if (!idm_lock)
+		{
+			// Enqueue the list update for when IDM becomes available
+			// Avoids a potential deadlock when a thread crashed while holding IDM mutex (it may crash the emulator later in Emu.Kill() etc, but GUI will work until then)
+			m_threads_created = -1;
+			break;
+		}
+
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/debugger_frame.cpp:884]: there's no reason to block signals before this section I think.
Also, all the GUI changes as well as the sceNpTrophy changes should better be seperate PRs.
This is way too crowded with random changes

Comment (elad335) [reply rpcs3/rpcs3qt/debugger_frame.cpp:884]: It's not because of QT, it's so idm mutex won't get locked for a while.

Comment (Megamouse) [reply rpcs3/rpcs3qt/debugger_frame.cpp:884]: This is all on the main thread, so I don't know where this would help

Comment (elad335) [reply rpcs3/rpcs3qt/debugger_frame.cpp:884]: Emulation threads use idm mutex all the time,

Comment (Megamouse) [reply rpcs3/rpcs3qt/debugger_frame.cpp:884]: The SignalBlocker is only meant to disable the signal/slot logic while the dropdown is cleared and refilled

Comment (elad335) [reply rpcs3/rpcs3qt/debugger_frame.cpp:884]: Again, not about QT at all. If this was a std::this_thread::sleep_for call for example I had move it out of the mutex scope as well for the same reason.

Comment (Megamouse) [reply rpcs3/rpcs3qt/debugger_frame.cpp:884]: what are you even talking about

Comment (elad335) [reply rpcs3/rpcs3qt/debugger_frame.cpp:884]: It's bad for performance to lock a mutex for a while.

**Pull request merged**